### PR TITLE
Asset Tag Helpers API Change

### PR DESCRIPTION
**Pull request by robwise**
- Change `env_javascript_include_tag` and `env_stylesheet_link_tag` API 
- rename `application_prod` -> `application_static`
- rename `application_dev` to `application_hot`
- pass `REACT_ON_RAILS_ENV='HOT'` environment variable in Procfile.dev

[Review on Reviewable](https://reviewable.io/reviews/shakacode/react-webpack-rails-tutorial/203)

### CODE REVIEW

File: Procfile.dev [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 # Development rails requires both rails and rails-assets
 # (and rails-server-assets if server rendering)
-rails: rails s
+rails: REACT_ON_RAILS_ENV='HOT' rails s
```
Comment (justin808) [inline Procfile.dev]: no quotes

File: app/assets/stylesheets/application_static.scss [hunk 2]
```diff
@@ -3,4 +3,4 @@
 @import 'app-bundle';
 
 // Non-webpack assets
-@import 'application_dev';
+@import 'application_hot';
```
Comment (justin808) [inline app/assets/stylesheets/application_static.scss]: this name might be confusing -- it means 

`application_non_webpack_js_files`: only the JS files that are not provided by webpack.

compare to 

`application_static`: All JS files, including webpack and non-webpack. This is used for production, testing, and running the development rails server without the environment of `REACT_ON_RAILS_ENV=HOT`

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -1,14 +1,24 @@
...
+  def env_stylesheet_link_tag(args = {})
+    send_tag_method(:stylesheet_link_tag, args)
+  end
+
+  private
+
+  def send_tag_method(tag_method_name, args)
+    asset_type = use_hot_reloading? ? :hot : :static
+    assets = Array(args[asset_type])
+    send(tag_method_name, *assets, args[:options])
```
Comment (justin808) [inline app/helpers/application_helper.rb]: remove the hot and static params from the options,

and let's change this so we don't have to have a special key of :options.

`env_javascript_include_tag(static: ['a', 'b'], turbo: true)

so just take out :static and :hot and pass the rest.

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -2,10 +2,9 @@
 <html>
 <head>
   <title>RailsReactTutorial</title>
-  <%= env_stylesheet_link_tag 'application_prod', 'application_dev', media: 'all', 'data-turbolinks-track' => true  %>
-  <%= env_javascript_include_tag nil, 'http://localhost:3500/vendor-bundle.js' %>
-  <%= env_javascript_include_tag nil, 'http://localhost:3500/app-bundle.js' %>
-  <%= env_javascript_include_tag 'application_prod', 'application_dev', 'data-turbolinks-track' => true %>
+  <%= env_stylesheet_link_tag(static: 'application_static', hot: 'application_hot', options: { media: 'all', 'data-turbolinks-track' => true })  %>
+  <%= env_javascript_include_tag(hot: ['http://localhost:3500/vendor-bundle.js', 'http://localhost:3500/app-bundle.js']) %>
+  <%= env_javascript_include_tag(static: 'application_static', hot: 'application_hot', options: { 'data-turbolinks-track' => true }) %>
```
Comment (justin808) [inline app/views/layouts/application.html.erb]: throw a comment in on why this is 2 lines, b/c we have different options for application_static and application_hot.

File: docs/jquery-with-react-on-rails.md [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 jQuery and jQuery-ujs are not required within `app/assets/javascript/application.js` and have been moved under`/client` and are managed by npm. The modules are exposed via entry point by [client/webpack.client.base.config.js](client/webpack.client.base.config.js) and, for `jquery-ujs`, in the [client/webpack.client.rails.build.config.js](client/webpack.client.rails.build.config.js) and the [client/webpack.client.rails.hot.config.js](client/webpack.client.rails.hot.config.js).
 
-In `application_dev.js` and `application_prod.js`, it's critical that any libraries that depend on jQuery come after the inclusion
+In `application_hot.js` and `application_static.js`, it's critical that any libraries that depend on jQuery come after the inclusion
```
Comment (justin808) [inline docs/jquery-with-react-on-rails.md]: `application_non_webpack_js.js`

**Comment by justin808** [comment]
We need updates to the Procfiles:

Procfile.dev --- everything
Procfile.hot -- hot assets rails server for dev mode
Procfile.rails -- static assets rails server for dev mode
Procfile.express
Procfile.spec -- only if not doing Procfile.rails -- builds the file for specs

**Comment by justin808** [comment]
Then we want to put the helpers in react_on_rails.

File: Procfile.dev (hunk 1 above)
Comment (robwise) [inline Procfile.dev]: fixed

File: app/helpers/application_helper.rb (hunk 3 above)
Comment (robwise) [inline app/helpers/application_helper.rb]: doine

File: app/views/layouts/application.html.erb (hunk 4 above)
Comment (robwise) [inline app/views/layouts/application.html.erb]: done

File: docs/jquery-with-react-on-rails.md (hunk 5 above)
Comment (robwise) [inline docs/jquery-with-react-on-rails.md]: done

File: app/assets/stylesheets/application_static.scss (hunk 2 above)
Comment (robwise) [inline app/assets/stylesheets/application_static.scss]: done

**Comment by robwise** [comment]
@justin808 I still need to add the colored warning message thing we discussed, then will be ready for merge.

**Comment by robwise** [comment]
@justin808 ready for review

**Comment by justin808** [comment]
Ready? :lgtm_strong: 

---

Reviewed 16 of 19 files at r1, 4 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[spec/support/ensure_assets_compiled.rb, line 8 \[r3\]](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/203#-K8p6LY4DzblYTxeVAKB:-K8p6LY5qGz0KlAUlR2o:-2084795023) ([raw file](https://github.com/shakacode/react-webpack-rails-tutorial/blob/2df524fa54ce95f30afe472ac86aec49b858f1b1/spec/support/ensure_assets_compiled.rb#L8)):*
@robwise when we move to this the generators, naturally, we don't put in the server part. We can check the config file REactOnRails to see if there's a server file. If you set to blank, that means no server rendering.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/203#-:-K8p6gOpEUvPUoX9s3_9:1770625613)*

**Comment by justin808** [comment]
Reviewed 16 of 19 files at r1, 4 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/203)*

**Pull request merged**